table/block.cc

```diff
@@ -377,27 +430,69 @@ bool IndexBlockIter::ParseNextIndexKey() {
     key_pinned_ = false;
   }
   value_ = Slice(p + non_shared, value_length);
-  while (restart_index_ + 1 < num_restarts_ &&
-         GetRestartPoint(restart_index_ + 1) < current_) {
-    ++restart_index_;
+  if (shared == 0) {
+    while (restart_index_ + 1 < num_restarts_ &&
+           GetRestartPoint(restart_index_ + 1) < current_) {
+      ++restart_index_;
+    }
+  }
+  // else we are in the middle of a restart interval and the restart_index_
+  // thus has not changed
+  if (value_delta_encoded_) {
+    assert(value_length == 0);
+    DecodeCurrentValue(shared);
   }
   return true;
 }
 
+// The format:
+// restart_point   0: k, v (off,sz), k, v (delta-sz), ..., k, v (delta-sz)
+// restart_point   1: k, v (off,sz), k, v (delta-sz), ..., k, v (delta-sz)
+// ...
+// restart_point n-1: k, v (off,sz), k, v (delta-sz), ..., k, v (delta-sz)
+// where, k is key, v is value, and its encoding is in parenthesis.
+// The format of each key is (shared_size, non_shared_size, shared, non_shared)
+// The format of each value, i.e., block hanlde, is (offset,size) whenever the
+// shared_size is 0, which included the first entry in each restart point.
+// Otherwise the format is delta-size = block handle size - size of last block
+// handle.
+void IndexBlockIter::DecodeCurrentValue(uint32_t shared) {
+  assert(value_delta_encoded_);
+  const char* l = value_.data() + 20;  // limit of two var-size uint64_t
```

Comment:
Why not use the real bound? This may in theory be out-of-bound?

Reply:
The bound is to protect against infinite loop in the case of corruption. Since this var-size encoded, the bound will be determined after the value is decoded.

Reply:
Yes but passing in an address that may be invalid still feels confusing to me. Why not pass a real bound, which I assume is `data_ + restarts_`?